src/components/tl/ucp/tl_ucp_context.c

```diff
@@ -0,0 +1,126 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2020.  ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#include "tl_ucp.h"
+#include "tl_ucp_tag.h"
+
+static void ucc_tl_ucp_req_init(void *request)
+{
+    ucc_tl_ucp_req_t *req = (ucc_tl_ucp_req_t *)request;
+    req->status           = UCC_OPERATION_INITIALIZED;
+}
+
+static void ucc_tl_ucp_req_cleanup(void *request)
+{
+}
+
+UCC_CLASS_INIT_FUNC(ucc_tl_ucp_context_t,
+                    const ucc_base_context_params_t *params,
+                    const ucc_base_config_t *config)
+{
+    ucc_tl_context_config_t *tl_config =
+        ucc_derived_of(config, ucc_tl_context_config_t);
+    ucc_status_t        ucc_status = UCC_OK;
+    ucp_worker_params_t worker_params;
+    ucp_worker_attr_t   worker_attr;
+    ucp_params_t        ucp_params;
+    ucp_config_t       *ucp_config;
+    ucp_context_h       ucp_context;
+    ucp_worker_h        ucp_worker;
+    ucs_status_t        status;
+
+    UCC_CLASS_CALL_SUPER_INIT(ucc_tl_context_t, tl_config->tl_lib);
+    status = ucp_config_read(params->prefix, NULL, &ucp_config);
+    if (UCS_OK != status) {
+        tl_error(&self->super, "failed to read ucp configuration, %s",
+                 ucs_status_string(status));
+        ucc_status = ucs_status_to_ucc_status(status);
+        goto err_cfg;
+    }
+
+    ucp_params.field_mask =
+        UCP_PARAM_FIELD_FEATURES | UCP_PARAM_FIELD_REQUEST_SIZE |
+        UCP_PARAM_FIELD_REQUEST_INIT | UCP_PARAM_FIELD_REQUEST_CLEANUP |
+        UCP_PARAM_FIELD_TAG_SENDER_MASK;
+    ucp_params.features        = UCP_FEATURE_TAG | UCP_FEATURE_RMA;
+    ucp_params.request_size    = sizeof(ucc_tl_ucp_req_t);
+    ucp_params.request_init    = ucc_tl_ucp_req_init;
+    ucp_params.request_cleanup = ucc_tl_ucp_req_cleanup;
+    ucp_params.tag_sender_mask = UCC_TL_UCP_TAG_SENDER_MASK;
+
+    if (params->estimated_num_ppn > 0) {
+        ucp_params.field_mask |= UCP_PARAM_FIELD_ESTIMATED_NUM_PPN;
+        ucp_params.estimated_num_ppn = params->estimated_num_ppn;
+    }
+
+    if (params->estimated_num_eps > 0) {
+        ucp_params.field_mask |= UCP_PARAM_FIELD_ESTIMATED_NUM_EPS;
+        ucp_params.estimated_num_eps = params->estimated_num_eps;
+    }
+
+    status = ucp_init(&ucp_params, ucp_config, &ucp_context);
```

Comment:
I would have expected ucp_init to be in the library init. We might create multiple context create and you don’t want to init that multiple times? I’m missing something.

Reply:
Is the behavior of params the same as UCC params? .i.e., it is more of a hints rather than constraints?

Reply:
Yes, you would think that ucp_init should go to lib_init BUT there reason it is in context is:
ucp_init is called after ucp_config_read. If you want to create ucp worker on a specific NET iface you need to read ucp config, then modify "UCX_NET_DEVICES" value and then create ucp_context + worker. 
This kind of means that in order to represent another device you need another "ucp_context". This is why it is here. (same in xccl)

regarding 2: yes params are very similar, rather hints